data/org.freedesktop.impl.portal.Notification.xml

```diff
@@ -37,10 +37,62 @@
 
         The ID can be used to later withdraw the notification.
         If the application reuses the same ID without withdrawing,
-        the notification is replaced by the new one.
+        the notification is updated with the new one. It's possible
+        to set ``show-as-new`` hint in the ``display-hint`` property
+        to animate replacing the notification instead of updating it.
 
         The format of the @notification is the same as for
-        :ref:`org.freedesktop.portal.Notification.AddNotification`.
+        :ref:`org.freedesktop.portal.Notification.AddNotification`,
+        expect for the following properties:
+
+        * ``icon`` (``v``)
+
+          A serialized icon to add to the notification. The format for
+          serialized icon is a tuple (sv) with the following supported keys:
+
+          * ``themed`` (``as``)
+
+            A themed icon containing an array of strings with the icon names.
+
+            This is the same format as a serialized `GThemedIcon
+            <https://docs.gtk.org/gio/class.ThemedIcon.html>`_.
+
+          * ``bytes`` (``ay``)
+
+            A bytes icon containing a bytestring with the icon data in
+            png, jpeg or svg form.
+
+            This is the same format as a serialized `GBytesIcon
+            <https://docs.gtk.org/gio/class.BytesIcon.html>`_.
+
+          * ``file`` (``s``)
+
+            A file icon containing a path to an image file in
+            png, jpeg or svg form.
+
+            This is the same format as a serialized `GFileIcon
+            <https://docs.gtk.org/gio/class.FileIcon.html>`_.
+
+        * ``sound`` (``v``)
+
+          A serialized sound to add to the notification.
+          Supported sound formats are ogg/opus, ogg/vorbis and wav/pcm.
+
+          The format for
+          serialized sound is a tuple (sv) with the following supported keys:
+
+          * ``bytes`` (``ay``)
+
+            A bytes sound containing a bytestring with the sound data.
+
+          * ``file`` (``s``)
+
+            A file sound containing an absolute path to a sound file.
+
+          To play the default sound the string ``default`` can be passed.
+          To play no sound at all the string ``silent`` can be passed.
```

Comment:
No, they are passed as a single `s` that's the reason why sound takes a `v` (and to stay consistent with icon)

Reply:
hmm, so essentially the type of sound action here is being encoded by the variant type of the `sound` key?

It feels to me like it would be a little clearer to describe, and to implement the code to handle, something more like: `sound` is an `sv` (rather than `v`). The `s` gives the type of sound action (`default`, `silent`, `bytes`, `file`), and then the `v` provides more information for that type (unit, unit, `ay` set of bytes, `ay` path).

That reminds me: technically you should use type `ay` (bytestring) for filenames/paths, rather than `s`, because the file system encoding may not be UTF-8 and/or files may contain invalid UTF-8.

Reply:
> It feels to me like it would be a little clearer to describe, and to implement the code to handle, something more like: sound is an sv (rather than v). The s gives the type of sound action (default, silent, bytes, file), and then the v provides more information for that type (unit, unit, ay set of bytes, ay path).

I would like that but we can't have an empty `v` for `default` and `silent`.

> That reminds me: technically you should use type ay (bytestring) for filenames/paths, rather than s, because the file system encoding may not be UTF-8 and/or files may contain invalid UTF-8.

Sounds reasonable  i didn't really think about it since GFileIcon uses `s` See: https://gitlab.gnome.org/GNOME/glib/-/blob/main/gio/gicon.c#L593 and
https://gitlab.gnome.org/GNOME/glib/-/blob/main/gio/gfileicon.c#L275

Reply:
> > It feels to me like it would be a little clearer to describe, and to implement the code to handle, something more like: sound is an sv (rather than v). The s gives the type of sound action (default, silent, bytes, file), and then the v provides more information for that type (unit, unit, ay set of bytes, ay path).
> 
> I would like that but we can't have an empty `v` for `default` and `silent`.

In pure `GVariant` you’d use `G_VARIANT_TYPE_UNIT` for this, but unfortunately that’s not supported by D-Bus, so the next best thing is to require passing an empty string or an empty array.

> > That reminds me: technically you should use type ay (bytestring) for filenames/paths, rather than s, because the file system encoding may not be UTF-8 and/or files may contain invalid UTF-8.
> 
> Sounds reasonable i didn't really think about it since GFileIcon uses `s` See: https://gitlab.gnome.org/GNOME/glib/-/blob/main/gio/gicon.c#L593 and https://gitlab.gnome.org/GNOME/glib/-/blob/main/gio/gfileicon.c#L275

oof, that’s an oversight. Looks like that serialisation format is going to have to change. I’ll look into that.

Reply:
> > > That reminds me: technically you should use type ay (bytestring) for filenames/paths, rather than s, because the file system encoding may not be UTF-8 and/or files may contain invalid UTF-8.
> > 
> > 
> > Sounds reasonable i didn't really think about it since GFileIcon uses `s` See: https://gitlab.gnome.org/GNOME/glib/-/blob/main/gio/gicon.c#L593 and https://gitlab.gnome.org/GNOME/glib/-/blob/main/gio/gfileicon.c#L275
> 
> oof, that’s an oversight. Looks like that serialisation format is going to have to change. I’ll look into that.

Wait, no, it’s correct in GLib because it’s serialising a `file:` URI, not a file path. Invalid UTF-8 will have been escaped by this point.

So as long as you make it clear in the docs that the string being passed around is a `file:` URI, rather than a plain path, it should be fine to continue using `s` rather than `ay`.

Reply:
ok nice, then i gonna change it to use URI to stay compatible.

Reply:
> In pure GVariant you’d use G_VARIANT_TYPE_UNIT for this, but unfortunately that’s not supported by D-Bus, so the next best thing is to require passing an empty string or an empty array.

Feels like having sound take a `v` is a little nicer. But no strong opinion.

Reply:
In my experience, having several levels of nesting gets pretty confusing for implementers pretty quickly. But up to you.